### DISCUSSION TITLE
Can epsilon in a medium can be changed during the simulation?

### DISCUSSION THREAD

**Discussion by manuelarturosantana** [NanoComp/meep#2654] [Q&A] [answered]
Hello,
I was wondering if epsilon in a medium can be changed during a simulation. I have tried something like
```
change_location = 3
def change_eps(sim):
    sim.geometry[change_location].material = mp.Medium(epsilon= 5 + np.cos(sim.meep_time()))
sim.run(change_eps, until=200)
```
This however doesn't seem to actually produce any change in epsilon during the simulation. 
Any advice gratefully accepted.

**Comment by 10UAikawa** [comment]
This is something I ran into while reading the MEEP documentition I dont know if this will help but this is what it said:

We are outputting the dielectric function ε but have wrapped its output function which would otherwise run at every time step in at_beginning, which does just what it says. There are [several other such functions](https://meep.readthedocs.io/en/latest/Python_User_Interface/#run-and-step-functions) to modify the output behavior — and you can, of course, write your own, and in fact you can do any computation or output you want at any time during the time evolution and even modify the simulation while it is running.

This is what I read in the tutorials/basics section in the MEEP documentation, it's the second tutorial "A 90° Bend".

**Reply by manuelarturosantana** [reply]
Thanks! I realized I had a typo in the sample code I provided, (now fixed), but yes, it should have a mp.at_every, and not just at the beginning.

**Reply by smartalecH** [reply, marked as answer]
You can also look at `set_materials`, which is intended to work well within arbitrary step functions. 

(Just search the docs for that function -- there's no active link).
